lib/posix-fdio/sendfile-shim.c

```diff
@@ -0,0 +1,277 @@
+	}
+#else /* !CONFIG_LIBVFSCORE */
+	return op ? SHIM_OFILE_OFILE_OFF : SHIM_OFILE_OFILE;
+#endif /* !CONFIG_LIBVFSCORE */
+}
+
+#if CONFIG_LIBVFSCORE
+
+/* HACK: Both uk_sys_ and vfscore_ ops share signatures, except for typeof f.
+ *
+ * Function type punning acceptable only because shim w/ vfscore; remove ASAP.
+ * Probably not compatible with very strictly-typed CFI either.
+ */
+typedef ssize_t (*sendfile_readf)(void *f, void *buf, size_t count);
+typedef ssize_t (*sendfile_preadf)(void *f, void *buf, size_t count, off_t off);
+typedef ssize_t (*sendfile_writef)(void *f, const void *buf, size_t count);
+typedef off_t (*sendfile_lseekf)(void *f, off_t offset, int whence);
+
+/* Arbitrary; small enough to fit on stacks, big enough to be efficient. */
+#define SENDFILE_BUFSZ 0x1000UL
+
+static
+ssize_t sendfile_compat(sendfile_readf readf, sendfile_writef writef,
+			sendfile_lseekf lseekf,
+			void *infile, void *outfile, size_t count)
+{
+	char tmp[SENDFILE_BUFSZ];
+	size_t sent = 0;
+	ssize_t r;
+
+	uk_pr_info("Fallback sendfile of 0x%zx bytes from stateful file offset\n",
+		   count);
+	do {
+		const size_t len = MIN(count, SENDFILE_BUFSZ);
+		size_t remain;
+		size_t written = 0;
+
+		r = readf(infile, tmp, len);
+		if (r <= 0)
+			break;
+		remain = r;
+		do {
+			r = writef(outfile, &tmp[written], remain);
+			if (r <= 0)
+				break;
+			remain -= r;
+			written += r;
+			sent += r;
+			count -= r;
+		} while (remain);
+		/* Rewind infile offset to before last read & break on error */
+		if (unlikely(remain)) {
+			off_t rc __maybe_unused;
+
+			rc = lseekf(infile, -remain, SEEK_CUR);
+			if (unlikely(rc < 0))
+				uk_pr_warn("Unable to revert input file as a result of sendfile write error %zd: %zd\n",
+					   r, rc);
+			break;
+		}
+	} while (count);
+	UK_ASSERT(sent || r <= 0); /* Return only error or nop through r */
+	return sent ? (ssize_t)sent : r;
+}
+
+static
+ssize_t psendfile_compat(sendfile_preadf preadf, sendfile_writef writef,
+			 void *infile, void *outfile, off_t *offset,
+			 size_t count)
+{
+	char tmp[SENDFILE_BUFSZ];
+	size_t sent = 0;
+	off_t off = *offset;
+	ssize_t r;
+
+	uk_pr_info("Fallback sendfile of 0x%zx bytes from offset 0x%zx\n",
+		   count, (size_t)off);
+	do {
+		const size_t len = MIN(count, SENDFILE_BUFSZ);
+		size_t remain;
+		size_t written = 0;
+
+		r = preadf(infile, tmp, len, off);
+		if (r <= 0)
+			break;
+		remain = r;
+		do {
+			r = writef(outfile, &tmp[written], remain);
+			if (r <= 0)
+				goto out;
+			remain -= r;
+			written += r;
+			sent += r;
+			count -= r;
+			off += r;
+		} while (remain);
+	} while (count);
+out:
+	*offset = off;
+	UK_ASSERT(sent || r <= 0); /* Return only error or nop through r */
+	return sent ? (ssize_t)sent : r;
+}
+
+#endif /* CONFIG_LIBVFSCORE */
+
+UK_SYSCALL_R_DEFINE(ssize_t, sendfile, int, out_fd, int, in_fd, off_t *, offp,
+		    size_t, count)
+{
+	union uk_shim_file out_sf;
+	union uk_shim_file in_sf;
+	int out_type;
+	int in_type;
+	ssize_t ret;
```

Comment:
So I assume the switch has all possible cases included, since `ret` is uninitialized and there is no `default`